File: Task2/shapes.py
class Shape:
    """Базовый класс для геометрических фигур."""
    def area(self):
        raise NotImplementedError("Метод area() должен быть переопределен")

    def perimeter(self):
        raise NotImplementedError("Метод perimeter() должен быть переопределен")

class Rectangle(Shape):
    """Производный класс: Прямоугольник."""
    def __init__(self, width, height):
        self.width = width
        self.height = height

    def area(self):
        return self.width * self.height

    def perimeter(self):
        return 2 * (self.width + self.height)

class Circle(Shape):
    """Производный класс: Круг."""
    def __init__(self, radius):
        self.radius = radius

    def area(self):
        return 3.1415 * self.radius ** 2

    def perimeter(self):
        return 2 * 3.1415 * self.radius

File: Task2/shapes.cpp
#include "shapes.h"
#include <stdexcept> // Для исключений

// Конструктор прямоугольника
Rectangle::Rectangle(double w, double h) : width(w), height(h) {
    if (w <= 0 || h <= 0) {
        throw std::invalid_argument("Ширина и высота должны быть положительными");
    }
}

// Расчет площади прямоугольника
double Rectangle::area() const { 
    return width * height; 
}

// Расчет периметра прямоугольника
double Rectangle::perimeter() const { 
    return 2 * (width + height); 
}

// Конструктор круга
Circle::Circle(double r) : radius(r) {
    if (r <= 0) {
        throw std::invalid_argument("Радиус должен быть положительным");
    }
}

// Расчет площади круга
double Circle::area() const { 
    return M_PI * radius * radius; 
}

// Расчет длины окружности (периметра)
double Circle::perimeter() const { 
    return 2 * M_PI * radius; 
}
